Two fixes to PSD estimation workflow

titodalcanton opened the pull request.

titodalcanton commented:
Ian, sounds like you have a test config you may want to run here. Otherwise I'll just repeat the C02 all-O1 estimate I did a while ago.

spxiwh commented on bin/hdfcoinc/pycbc_make_psd_estimation_workflow:
```diff
@@ -109,7 +109,8 @@ for ifo, segments in analyzable_segs.items():
 
-            if len(partial_broken_segments) >= max_segs_per_job:
+            if (len(partial_broken_segments) >= max_segs_per_job) \
+                    or (start_time + psd_job_length + pad_data > seg[1]):
```
What exactly is the intention here in terms of parallelization? Previously I had understood that each job would analyse max_segs_per_job * psd_job_length. Now it seems that at the end of every segment a new job will be created. So previously all jobs would have the same run time, but now there will be quite a lot of inbalance. A lot of jobs looking at short analysis segments will run very quickly (potentially containing 1 or a small number of "broken segments"). This will also mean a lot more PSD jobs will be created, potentially leading to some I/O load when all of the jobs run at the same time .... I think the previous approach, where all jobs have the same run time was better!

However, I can see that the *last* group of broken segments would not have been analysed as there was no code to create a final job to analyse the < max_segs_per_job segments at the end!

Also is it possible to change "broken" -> "split". "Broken" implies not working and worried me when I saw this in an output file name!

spxiwh commented:
Hi Tito, configuring the max-segments job in the .ini file makes sense. I'm confused by the other change though, as described in my in-line comment above.

Also please rebase to get landscape working.

titodalcanton commented:
As you correctly noted, previously the last group of split segments would not have been analysed (I went to look and realized it after your question on Mattermost). Although that's probably going to have a small effect on the PSD estimate, I think that's not the intended behavior, so I fixed it.

Another way of fixing it would be to extend the last job to include the final split segments. This would make some of the jobs a bit longer instead of creating additional shorter jobs. Would you be happier with that in terms of job run time inbalance?

spxiwh commented:
In terms of job balance, I think it would make more sense to have one *additional* job (instead of extending the last job) which contains only the final segments. That then gives *one* short job instead of many.

titodalcanton commented:
Ian, I rewrote that part according to your comments and hopefully made it a bit clearer. I also added the missing [merge_psds] section to the documentation. I'm testing this on the C02 O1 data at

https://www.atlas.aei.uni-hannover.de/~tito/LSC/o1/psd-estim-test3/

spxiwh commented:
This all looks good, thanks Tito. Once test completes I will merge (please post when test is completed)

titodalcanton commented:
Test results:
https://www.atlas.aei.uni-hannover.de/~tito/LSC/o1/psd-estim-test3/H1L1-PLOT_SPECTRUM-1126051217-11203200.png

Baseline:
https://www.atlas.aei.uni-hannover.de/~tito/LSC/o1/o1-psd-estim-c02/H1L1-PLOT_SPECTRUM-1126051217-11203200.png

The relative difference between the two average PSD series is within 0.04% over the whole frequency range (within 0.004% above 10 Hz).
